File: Source/Tortuga/Graphics/Vulkan/DescriptorSet.cpp
#include "./DescriptorSet.hpp"

namespace Tortuga
{
namespace Graphics
{
namespace Vulkan
{
namespace DescriptorSet
{
DescriptorSet Create(Device::Device device, DescriptorPool::DescriptorPool pool, DescriptorLayout::DescriptorLayout layout, uint32_t setArrayCount)
{
  DescriptorSet data = {};
  data.Device = device.Device;
  data.Pool = pool;
  data.Layout = layout;
  data.SetArrayCount = setArrayCount;

  VkDescriptorSetAllocateInfo info = {};
  {
    info.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_ALLOCATE_INFO;
    info.descriptorPool = pool.Pool,
    info.descriptorSetCount = setArrayCount;
    info.pSetLayouts = &layout.Layouts;
  }
  ErrorCheck::Callback(vkAllocateDescriptorSets(device.Device, &info, &data.set));
  return data;
}
void UpdateDescriptorSet(DescriptorSet data, std::vector<Buffer::Buffer> content, uint32_t setArrayIndex)
{
  if (data.Layout.BindingsAmount != content.size())
  {
    Console::Error("provided Content does not match this descriptor set size");
    return;
  }

  std::vector<VkDescriptorBufferInfo> bufferInfos(content.size());
  std::vector<VkWriteDescriptorSet> writeInfos(data.Layout.BindingsAmount);
  for (uint32_t i = 0; i < data.Layout.BindingsAmount; i++)
  {
    {
      bufferInfos[i] = {};
      bufferInfos[i].buffer = content[i].Buffer;
      bufferInfos[i].offset = 0;
      bufferInfos[i].range = content[i].Size;
    }

    writeInfos[i].sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
    writeInfos[i].dstSet = data.set;
    writeInfos[i].dstBinding = i;
    writeInfos[i].dstArrayElement = setArrayIndex;
    writeInfos[i].descriptorCount = data.SetArrayCount;
    writeInfos[i].descriptorType = data.Layout.Types[i];

    writeInfos[i].pBufferInfo = &(bufferInfos[i]);
    writeInfos[i].pImageInfo = VK_NULL_HANDLE;
    writeInfos[i].pTexelBufferView = VK_NULL_HANDLE;
  }
  vkUpdateDescriptorSets(data.Device, writeInfos.size(), writeInfos.data(), 0, 0);
}

void UpdateDescriptorSet(DescriptorSet data, std::vector<ImageView::ImageView> content, Sampler::Sampler sampler, uint32_t setArrayIndex)
{
  if (data.Layout.BindingsAmount != content.size())
  {
    Console::Error("provided Content does not match this descriptor set size");
    return;
  }

  std::vector<VkDescriptorImageInfo> imageInfo(content.size());
  std::vector<VkWriteDescriptorSet> writeInfos(data.Layout.BindingsAmount);
  for (uint32_t i = 0; i < data.Layout.BindingsAmount; i++)
  {
    {
      imageInfo[i] = {};
      imageInfo[i].imageLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
      imageInfo[i].imageView = content[i].View;
      imageInfo[i].sampler = sampler.Sampler;
    }

    writeInfos[i].sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
    writeInfos[i].dstSet = data.set;
    writeInfos[i].dstBinding = i;
    writeInfos[i].dstArrayElement = setArrayIndex;
    writeInfos[i].descriptorCount = data.SetArrayCount;
    writeInfos[i].descriptorType = data.Layout.Types[i];

    writeInfos[i].pBufferInfo = VK_NULL_HANDLE;
    writeInfos[i].pImageInfo = &(imageInfo[i]);
    writeInfos[i].pTexelBufferView = VK_NULL_HANDLE;
  }
  vkUpdateDescriptorSets(data.Device, writeInfos.size(), writeInfos.data(), 0, 0);
}
} // namespace DescriptorSet
} // namespace Vulkan
} // namespace Graphics
} // namespace Tortuga

File: Tortuga.Graphics/API/DescriptorSet.cs
#pragma warning disable CS1591
using System;
using System.Collections.Generic;
using Tortuga.Utils;
using Vulkan;

namespace Tortuga.Graphics.API
{
    public class DescriptorSet
    {
        public DescriptorPool DescriptorPool => _descriptorPool;
        public Device Device => _device;
        public VkDescriptorSet Handle => _handle;

        private DescriptorPool _descriptorPool;
        private Device _device;
        private VkDescriptorSet _handle;

        public unsafe DescriptorSet(DescriptorPool descriptorPool, uint setCount = 1)
        {
            _device = descriptorPool.Device;
            _descriptorPool = descriptorPool;

            var layouts = new NativeList<VkDescriptorSetLayout>(setCount);
            for (int i = 0; i < setCount; i++)
                layouts.Add(descriptorPool.Layout.Handle);

            var allocateInfo = new VkDescriptorSetAllocateInfo
            {
                sType = VkStructureType.DescriptorSetAllocateInfo,
                descriptorPool = descriptorPool.Handle,
                descriptorSetCount = setCount,
                pSetLayouts = (VkDescriptorSetLayout*)layouts.Data.ToPointer()
            };

            VkDescriptorSet descriptorSet;
            if (VulkanNative.vkAllocateDescriptorSets(
                _device.Handle,
                &allocateInfo,
                &descriptorSet
            ) != VkResult.Success)
                throw new Exception("failed to allocate descriptor sets");
            _handle = descriptorSet;
        }

        public unsafe void UpdateBuffers(List<Buffer> buffers)
        {
            if (buffers.Count != _descriptorPool.Layout.Bindings.Count)
                throw new InvalidOperationException("buffers length should match descriptor layout bindings");

            var descriptorSetBufferInfos = new NativeList<VkDescriptorBufferInfo>();
            var writeDescriptorSets = new NativeList<VkWriteDescriptorSet>();

            for (int i = 0; i < _descriptorPool.Layout.Bindings.Count; i++)
            {
                var binding = _descriptorPool.Layout.Bindings[i];
                var descriptorSetBufferInfo = new VkDescriptorBufferInfo
                {
                    buffer = buffers[i].Handle,
                    offset = 0,
                    range = buffers[i].Size
                };

                var writeDescriptorSet = new VkWriteDescriptorSet
                {
                    sType = VkStructureType.WriteDescriptorSet,
                    dstSet = _handle,
                    dstBinding = binding.Index,
                    dstArrayElement = 0,
                    descriptorCount = binding.DescriptorCounts,
                    descriptorType = binding.DescriptorType,
                    pBufferInfo = &descriptorSetBufferInfo,
                };
                descriptorSetBufferInfos.Add(descriptorSetBufferInfo);
                writeDescriptorSets.Add(writeDescriptorSet);
            }
            VulkanNative.vkUpdateDescriptorSets(
                _device.Handle,
                writeDescriptorSets.Count,
                (VkWriteDescriptorSet*)writeDescriptorSets.Data.ToPointer(),
                0,
                null
            );
        }

        public unsafe void UpdateBuffer(Buffer buffer, int binding)
        {
            var vulkanBinding = _descriptorPool.Layout.Bindings[binding];
            var descriptorSetBufferInfo = new VkDescriptorBufferInfo
            {
                buffer = buffer.Handle,
                offset = 0,
                range = buffer.Size
            };

            var writeDescriptorSet = new VkWriteDescriptorSet
            {
                sType = VkStructureType.WriteDescriptorSet,
                dstSet = _handle,
                dstBinding = vulkanBinding.Index,
                dstArrayElement = 0,
                descriptorCount = vulkanBinding.DescriptorCounts,
                descriptorType = vulkanBinding.DescriptorType,
                pBufferInfo = &descriptorSetBufferInfo,
            };
            VulkanNative.vkUpdateDescriptorSets(
                _device.Handle,
                1,
                &writeDescriptorSet,
                0,
                null
            );
        }

        public unsafe void UpdateSampledImages(List<ImageView> views, List<Sampler> samplers)
        {
            if (views.Count != _descriptorPool.Layout.Bindings.Count)
                throw new InvalidOperationException("views length should match descriptor layout bindings");
            if (samplers.Count != _descriptorPool.Layout.Bindings.Count)
                throw new InvalidOperationException("samplers length should match descriptor layout bindings");

            var descriptorImageInfos = new NativeList<VkDescriptorImageInfo>();
            var writeDescriptorSets = new NativeList<VkWriteDescriptorSet>();

            for (int i = 0; i < _descriptorPool.Layout.Bindings.Count; i++)
            {
                var binding = _descriptorPool.Layout.Bindings[i];
                var descriptorImageInfo = new VkDescriptorImageInfo
                {
                    imageLayout = views[i].Image.Layout,
                    imageView = views[i].Handle,
                    sampler = samplers[i].Handle
                };

                var writeDescriptorSet = new VkWriteDescriptorSet
                {
                    sType = VkStructureType.WriteDescriptorSet,
                    dstSet = _handle,
                    dstBinding = binding.Index,
                    dstArrayElement = 0,
                    descriptorCount = binding.DescriptorCounts,
                    descriptorType = binding.DescriptorType,
                    pImageInfo = &descriptorImageInfo
                };
                descriptorImageInfos.Add(descriptorImageInfo);
                writeDescriptorSets.Add(writeDescriptorSet);
            }
            VulkanNative.vkUpdateDescriptorSets(
                _device.Handle,
                writeDescriptorSets.Count,
                (VkWriteDescriptorSet*)writeDescriptorSets.Data.ToPointer(),
                0,
                null
            );
        }

        public unsafe void UpdateSampledImage(ImageView view, Sampler sampler, int binding)
        {
            var vulkanBinding = _descriptorPool.Layout.Bindings[binding];
            var descriptorImageInfo = new VkDescriptorImageInfo
            {
                imageLayout = view.Image.Layout,
                imageView = view.Handle,
                sampler = sampler.Handle
            };

            var writeDescriptorSet = new VkWriteDescriptorSet
            {
                sType = VkStructureType.WriteDescriptorSet,
                dstSet = _handle,
                dstBinding = vulkanBinding.Index,
                dstArrayElement = 0,
                descriptorCount = vulkanBinding.DescriptorCounts,
                descriptorType = vulkanBinding.DescriptorType,
                pImageInfo = &descriptorImageInfo
            };
            VulkanNative.vkUpdateDescriptorSets(
                _device.Handle,
                1,
                &writeDescriptorSet,
                0,
                null
            );
        }
    }
}